File: back/realtime_api.py
# realtime_api.py - Real-time API service for frontend integration

import requests
import json
from typing import Dict, List, Optional

class InterviewPlatformAPI:
    def __init__(self, base_url: str = "http://localhost:5000"):
        self.base_url = base_url
        self.token = None
        self.headers = {"Content-Type": "application/json"}
    
    def _update_headers(self):
        """Update headers with authentication token"""
        if self.token:
            self.headers["Authorization"] = f"Bearer {self.token}"
    
    def _make_request(self, method: str, endpoint: str, data: Optional[Dict] = None, files: Optional[Dict] = None):
        """Make HTTP request with error handling"""
        url = f"{self.base_url}{endpoint}"
        
        try:
            if method.upper() == "GET":
                response = requests.get(url, headers=self.headers)
            elif method.upper() == "POST":
                if files:
                    # Remove Content-Type for file uploads
                    headers = {k: v for k, v in self.headers.items() if k != "Content-Type"}
                    response = requests.post(url, headers=headers, files=files, data=data)
                else:
                    response = requests.post(url, headers=self.headers, json=data)
            elif method.upper() == "PUT":
                response = requests.put(url, headers=self.headers, json=data)
            elif method.upper() == "DELETE":
                response = requests.delete(url, headers=self.headers)
            else:
                raise ValueError(f"Unsupported HTTP method: {method}")
            
            return response.json() if response.content else {}
        
        except requests.exceptions.RequestException as e:
            return {"success": False, "error": str(e)}
        except json.JSONDecodeError:
            return {"success": False, "error": "Invalid JSON response"}
    
    # Authentication Methods
    def login(self, email: str, password: str, role: str = None) -> Dict:
        """User login"""
        data = {"email": email, "password": password}
        if role:
            data["role"] = role
        
        response = self._make_request("POST", "/auth/login", data)
        
        if response.get("success") and "access_token" in response:
            self.token = response["access_token"]
            self._update_headers()
        
        return response
    
    def register(self, user_data: Dict) -> Dict:
        """User registration"""
        response = self._make_request("POST", "/auth/register", user_data)
        
        if response.get("success") and "access_token" in response:
            self.token = response["access_token"]
            self._update_headers()
        
        return response
    
    def get_profile(self) -> Dict:
        """Get user profile"""
        return self._make_request("GET", "/auth/profile")
    
    def update_profile(self, user_data: Dict) -> Dict:
        """Update user profile"""
        return self._make_request("PUT", "/auth/profile", user_data)
    
    def logout(self):
        """Logout user"""
        self.token = None
        self.headers = {"Content-Type": "application/json"}
    
    # Job Management Methods
    def get_jobs(self) -> Dict:
        """Get all jobs"""
        return self._make_request("GET", "/jobs")
    
    def get_job_by_id(self, job_id: str) -> Dict:
        """Get job by ID"""
        return self._make_request("GET", f"/jobs/{job_id}")
    
    def create_job(self, job_data: Dict) -> Dict:
        """Create new job (Admin only)"""
        return self._make_request("POST", "/jobs", job_data)
    
    def update_job(self, job_id: str, job_data: Dict) -> Dict:
        """Update job (Admin only)"""
        return self._make_request("PUT", f"/jobs/{job_id}", job_data)
    
    def delete_job(self, job_id: str) -> Dict:
        """Delete job (Admin only)"""
        return self._make_request("DELETE", f"/jobs/{job_id}")
    
    # Interview Management Methods
    def get_interviews(self) -> Dict:
        """Get user's interviews"""
        return self._make_request("GET", "/interviews")
    
    def schedule_interview(self, job_id: str, scheduled_at: str) -> Dict:
        """Schedule new interview"""
        data = {
            "job_id": job_id,
            "scheduled_at": scheduled_at
        }
        return self._make_request("POST", "/interviews", data)
    
    def get_interview_by_id(self, interview_id: str) -> Dict:
        """Get interview by ID"""
        return self._make_request("GET", f"/interviews/{interview_id}")
    
    def add_chat_message(self, interview_id: str, sender: str, message: str) -> Dict:
        """Add message to interview transcript"""
        data = {
            "sender": sender,
            "message": message
        }
        return self._make_request("POST", f"/interviews/{interview_id}/messages", data)
    
    def complete_interview(self, interview_id: str, ai_summary: str, performance_score: float, feedback: str) -> Dict:
        """Complete interview with results"""
        data = {
            "ai_summary": ai_summary,
            "performance_score": performance_score,
            "feedback": feedback
        }
        return self._make_request("POST", f"/interviews/{interview_id}/complete", data)
    
    # Resume and Question Methods
    def upload_resume(self, file_path: str) -> Dict:
        """Upload resume and generate questions"""
        with open(file_path, 'rb') as f:
            files = {"resume": f}
            response = self._make_request("POST", "/uploadcv", files=files)
        return response
    
    def submit_answers(self, answers: List[Dict]) -> Dict:
        """Submit interview answers for evaluation"""
        return self._make_request("POST", "/submitqna", answers)
    
    def get_interview_summary(self, performance_score: float, ai_summary: str) -> Dict:
        """Get interview summary"""
        data = {
            "performanceScore": performance_score,
            "aiSummary": ai_summary
        }
        return self._make_request("POST", "/interview-summary", data)
    
    # Activity Methods
    def get_activities(self) -> Dict:
        """Get user activities"""
        return self._make_request("GET", "/activities")
    
    # Mock AI Response (for chat simulation)
    def generate_ai_response(self, user_message: str, context: List[Dict] = None) -> Dict:
        """Generate AI response for chat interview"""
        # This could be enhanced to call a real AI service
        # For now, return a simple response
        responses = [
            "That's an interesting point. Can you elaborate on that?",
            "Great! Now, let me ask you about your experience with...",
            "I see. How would you handle a situation where...",
            "Excellent answer! Moving on to the next question...",
            "Can you walk me through your thought process on that?",
            "That's a good approach. What challenges did you face when implementing that?",
            "I understand. What would you do differently next time?",
            "Perfect! Let's discuss your experience with team collaboration.",
            "Interesting! How do you stay updated with the latest technologies?",
            "Thank you for that detailed explanation. Let's move forward."
        ]
        
        import random
        response_message = random.choice(responses)
        
        return {
            "success": True,
            "message": response_message
        }

# Example usage
if __name__ == "__main__":
    # Initialize API client
    api = InterviewPlatformAPI()
    
    # Test login
    login_response = api.login("john@example.com", "candidate123", "candidate")
    print("Login Response:", login_response)
    
    if login_response.get("success"):
        # Test getting jobs
        jobs_response = api.get_jobs()
        print("Jobs Response:", jobs_response)
        
        # Test getting profile
        profile_response = api.get_profile()
        print("Profile Response:", profile_response)
        
        # Test getting interviews
        interviews_response = api.get_interviews()
        print("Interviews Response:", interviews_response)
        
        # Test getting activities
        activities_response = api.get_activities()
        print("Activities Response:", activities_response)

File: front-bolt/src/services/realTimeApi.ts
// realTimeApi.ts - Real-time API service for MongoDB integration

import { User, Job, Interview, ChatMessage } from '../types';

// API Configuration
const API_BASE_URL = 'http://localhost:5000';

class APIClient {
  private token: string | null = null;

  constructor() {
    // Try to get token from localStorage
    this.token = localStorage.getItem('access_token');
  }

  private getHeaders(): HeadersInit {
    const headers: HeadersInit = {
      'Content-Type': 'application/json',
    };

    if (this.token) {
      headers['Authorization'] = `Bearer ${this.token}`;
    }

    return headers;
  }

  private async request<T>(
    endpoint: string,
    options: RequestInit = {}
  ): Promise<T> {
    const url = `${API_BASE_URL}${endpoint}`;
    
    const config: RequestInit = {
      ...options,
      headers: {
        ...this.getHeaders(),
        ...options.headers,
      },
    };

    try {
      const response = await fetch(url, config);
      const data = await response.json();
      
      if (!response.ok) {
        throw new Error(data.error || `HTTP error! status: ${response.status}`);
      }

      return data;
    } catch (error) {
      console.error(`API request failed: ${url}`, error);
      throw error;
    }
  }

  private setToken(token: string) {
    this.token = token;
    localStorage.setItem('access_token', token);
  }

  private clearToken() {
    this.token = null;
    localStorage.removeItem('access_token');
  }

  // Authentication methods
  async login(email: string, password: string, role: 'candidate' | 'admin') {
    try {
      const response = await this.request<{
        success: boolean;
        user: User;
        access_token: string;
      }>('/auth/login', {
        method: 'POST',
        body: JSON.stringify({ email, password, role }),
      });

      if (response.success && response.access_token) {
        this.setToken(response.access_token);
        return { success: true, user: response.user, token: response.access_token };
      }

      return { success: false, error: 'Login failed' };
    } catch (error) {
      return { success: false, error: error instanceof Error ? error.message : 'Login failed' };
    }
  }

  async signup(userData: Partial<User>, password: string) {
    try {
      const response = await this.request<{
        success: boolean;
        user: User;
        access_token: string;
      }>('/auth/register', {
        method: 'POST',
        body: JSON.stringify({ ...userData, password }),
      });

      if (response.success && response.access_token) {
        this.setToken(response.access_token);
        return { success: true, user: response.user, token: response.access_token };
      }

      return { success: false, error: 'Registration failed' };
    } catch (error) {
      return { success: false, error: error instanceof Error ? error.message : 'Registration failed' };
    }
  }

  async updateProfile(userData: Partial<User>) {
    try {
      const response = await this.request<{
        success: boolean;
        user: User;
      }>('/auth/profile', {
        method: 'PUT',
        body: JSON.stringify(userData),      });

      return response;
    } catch (error) {
      return { success: false, error: error instanceof Error ? error.message : 'Profile update failed' };
    }
  }

  async getCurrentUser() {
    try {
      const response = await this.request<{
        success: boolean;
        user: User;
      }>('/auth/profile', {
        method: 'GET',
      });

      return response;
    } catch (error) {
      return { success: false, error: error instanceof Error ? error.message : 'Failed to get current user' };
    }
  }

  logout() {
    this.clearToken();
  }

  // Job methods
  async getJobs() {
    try {
      const response = await this.request<{
        success: boolean;
        jobs: Job[];
      }>('/jobs');

      return response;
    } catch (error) {
      return { success: false, error: error instanceof Error ? error.message : 'Failed to fetch jobs' };
    }
  }

  async getJobById(id: string) {
    try {
      const response = await this.request<{
        success: boolean;
        job: Job;
      }>(`/jobs/${id}`);

      return response;
    } catch (error) {
      return { success: false, error: error instanceof Error ? error.message : 'Failed to fetch job' };
    }
  }

  async createJob(jobData: Omit<Job, 'id' | 'createdAt'>) {
    try {
      const response = await this.request<{
        success: boolean;
        job: Job;
      }>('/jobs', {
        method: 'POST',
        body: JSON.stringify(jobData),
      });

      return response;
    } catch (error) {
      return { success: false, error: error instanceof Error ? error.message : 'Failed to create job' };
    }
  }

  async updateJob(id: string, jobData: Partial<Job>) {
    try {
      const response = await this.request<{
        success: boolean;
        job: Job;
      }>(`/jobs/${id}`, {
        method: 'PUT',
        body: JSON.stringify(jobData),
      });

      return response;
    } catch (error) {
      return { success: false, error: error instanceof Error ? error.message : 'Failed to update job' };
    }
  }

  async deleteJob(id: string) {
    try {
      const response = await this.request<{
        success: boolean;
      }>(`/jobs/${id}`, {
        method: 'DELETE',
      });

      return response;
    } catch (error) {
      return { success: false, error: error instanceof Error ? error.message : 'Failed to delete job' };
    }
  }

  // Interview methods
  async getInterviews(candidateId?: string) {
    try {
      const response = await this.request<{
        success: boolean;
        interviews: Interview[];
      }>('/interviews');

      return response;
    } catch (error) {
      return { success: false, error: error instanceof Error ? error.message : 'Failed to fetch interviews' };
    }
  }

  async scheduleInterview(candidateId: string, jobId: string, scheduledAt: Date) {
    try {
      const response = await this.request<{
        success: boolean;
        interview: Interview;
      }>('/interviews', {
        method: 'POST',
        body: JSON.stringify({
          job_id: jobId,
          scheduled_at: scheduledAt.toISOString(),
        }),
      });

      return response;
    } catch (error) {
      return { success: false, error: error instanceof Error ? error.message : 'Failed to schedule interview' };
    }
  }

  async getInterviewById(id: string) {
    try {
      const response = await this.request<{
        success: boolean;
        interview: Interview;
      }>(`/interviews/${id}`);

      return response;
    } catch (error) {
      return { success: false, error: error instanceof Error ? error.message : 'Failed to fetch interview' };
    }
  }

  async addChatMessage(interviewId: string, message: Omit<ChatMessage, 'id' | 'timestamp'>) {
    try {
      const response = await this.request<{
        success: boolean;
        message: ChatMessage;
      }>(`/interviews/${interviewId}/messages`, {
        method: 'POST',
        body: JSON.stringify(message),
      });

      return response;
    } catch (error) {
      return { success: false, error: error instanceof Error ? error.message : 'Failed to add message' };
    }
  }

  async completeInterview(interviewId: string, aiSummary: string, performanceScore: number, feedback: string) {
    try {
      const response = await this.request<{
        success: boolean;
        message: string;
      }>(`/interviews/${interviewId}/complete`, {
        method: 'POST',
        body: JSON.stringify({
          ai_summary: aiSummary,
          performance_score: performanceScore,
          feedback: feedback,
        }),
      });

      return response;
    } catch (error) {
      return { success: false, error: error instanceof Error ? error.message : 'Failed to complete interview' };
    }
  }

  // Generate AI response for chat
  async generateAIResponse(userMessage: string, context: ChatMessage[]) {
    // This is a simple mock implementation
    // In a real application, you would send this to your AI service
    await new Promise(resolve => setTimeout(resolve, 1000 + Math.random() * 2000));

    const responses = [
      "That's an interesting point. Can you elaborate on your experience with that technology?",
      "Great! Can you walk me through your problem-solving approach for that scenario?",
      "I see. How do you handle challenges when working in a team environment?",
      "Excellent answer! Now, let me ask you about your experience with...",
      "Can you provide a specific example of when you implemented that solution?",
      "That's a good approach. What would you do if you encountered performance issues?",
      "Interesting! How do you stay updated with the latest industry trends?",
      "Perfect! Let's discuss your experience with project management and deadlines.",
      "Thank you for that detailed explanation. How do you handle code reviews?",
      "Great insight! What's your approach to testing and quality assurance?"
    ];

    const randomResponse = responses[Math.floor(Math.random() * responses.length)];

    return {
      success: true,
      message: randomResponse
    };
  }

  // Activity methods
  async getActivities() {
    try {
      const response = await this.request<{
        success: boolean;
        activities: any[];
      }>('/activities');

      return response;
    } catch (error) {
      return { success: false, error: error instanceof Error ? error.message : 'Failed to fetch activities' };
    }
  }

  // Resume upload
  async uploadResume(file: File) {
    try {
      const formData = new FormData();
      formData.append('resume', file);

      const response = await fetch(`${API_BASE_URL}/uploadcv`, {
        method: 'POST',
        headers: {
          'Authorization': `Bearer ${this.token}`,
        },
        body: formData,
      });

      const data = await response.json();

      if (!response.ok) {
        throw new Error(data.error || 'Upload failed');
      }

      return data;
    } catch (error) {
      return { success: false, error: error instanceof Error ? error.message : 'Upload failed' };
    }
  }

  async submitAnswers(answers: any[]) {
    try {
      const response = await this.request<{
        success: boolean;
        results: any[];
      }>('/submitqna', {
        method: 'POST',
        body: JSON.stringify(answers),
      });

      return response;
    } catch (error) {
      return { success: false, error: error instanceof Error ? error.message : 'Failed to submit answers' };
    }
  }
}

// Create singleton instance
const apiClient = new APIClient();

// Export services that match the existing mockApi interface
export const mockAuthService = {
  async login(email: string, password: string, role: 'candidate' | 'admin') {
    return apiClient.login(email, password, role);
  },

  async signup(userData: Partial<User>, password: string) {
    return apiClient.signup(userData, password);
  },

  async updateProfile(userData: Partial<User>) {
    return apiClient.updateProfile(userData);
  },

  async getCurrentUser() {
    return apiClient.getCurrentUser();
  },

  logout() {
    apiClient.logout();
  }
};

export const mockJobService = {
  async getJobs() {
    return apiClient.getJobs();
  },

  async getJobById(id: string) {
    return apiClient.getJobById(id);
  },

  async createJob(jobData: Omit<Job, 'id' | 'createdAt'>) {
    return apiClient.createJob(jobData);
  },

  async updateJob(id: string, jobData: Partial<Job>) {
    return apiClient.updateJob(id, jobData);
  },

  async deleteJob(id: string) {
    return apiClient.deleteJob(id);
  }
};

export const mockInterviewService = {
  async getInterviews(candidateId?: string) {
    return apiClient.getInterviews(candidateId);
  },

  async scheduleInterview(candidateId: string, jobId: string, scheduledAt: Date) {
    return apiClient.scheduleInterview(candidateId, jobId, scheduledAt);
  },

  async getInterviewById(id: string) {
    return apiClient.getInterviewById(id);
  },

  async addChatMessage(interviewId: string, message: Omit<ChatMessage, 'id' | 'timestamp'>) {
    return apiClient.addChatMessage(interviewId, message);
  },

  async completeInterview(interviewId: string, aiSummary: string, performanceScore: number, feedback: string) {
    return apiClient.completeInterview(interviewId, aiSummary, performanceScore, feedback);
  },

  async generateAIResponse(userMessage: string, context: ChatMessage[]) {
    return apiClient.generateAIResponse(userMessage, context);
  }
};

export const mockAdminService = {
  async getCandidates() {
    // This would need a specific endpoint in the backend
    return { success: true, candidates: [] };
  },

  async getInterviewAnalytics() {
    const activities = await apiClient.getActivities();
    return {
      success: true,
      analytics: {
        totalInterviews: 0,
        completedInterviews: 0,
        averageScore: 0,
        highPerformers: 0,
        recentInterviews: [],
      }
    };
  }
};

export default apiClient;
